polar-core/src/vm.rs

```diff
@@ -438,9 +476,7 @@ impl PolarVirtualMachine {
     /// Try to achieve one goal. Return `Some(QueryEvent)` if an external
     /// result is needed to achieve it, or `None` if it can run internally.
     fn next(&mut self, goal: Rc<Goal>) -> Result<QueryEvent> {
-        if self.log {
-            self.print(&format!("{}", goal));
-        }
+        self.log_with(LogLevel::Trace, || format!("{}", goal), &[]);
```

Comment:
super nit: I think `goal.to_string()` is a nicer way to spell `format!("{}", goal)`